lib/internal/fs/streams.js

```diff
@@ -180,13 +184,17 @@ ReadStream.prototype._read = function(n) {
       this.emit('error', er);
     } else {
       let b = null;
+      let alignedOffset = 0;
       // Now that we know how much data we have actually read, re-wind the
       // 'used' field if we can, and otherwise allow the remainder of our
       // reservation to be used as a new pool later.
-      if (start + toRead === thisPool.used && thisPool === pool)
-        thisPool.used += bytesRead - toRead;
-      else if (toRead - bytesRead > kMinPoolSpace)
-        poolFragments.push(thisPool.slice(start + bytesRead, start + toRead));
+      if (start + toRead === thisPool.used && thisPool === pool) {
+        alignedOffset = thisPool.used + bytesRead - toRead;
+        thisPool.used = roundUpToMultipleOf8(alignedOffset);
+      } else if (toRead - bytesRead > kMinPoolSpace) {
+        alignedOffset = roundUpToMultipleOf8(start + bytesRead);
+        poolFragments.push(thisPool.slice(alignedOffset, start + toRead));
```

Comment:
If you're only using `alignedOffset` inside the if/else arms, can you make it `const` and scope it to inside the arms?

Is `start + toRead` correct when `alignedOffset` may not be `start + bytesRead`? What I mean is, shouldn't `start + toRead` be rounded up too?

Reply:
Rounding up is unsafe here, since `toRead` is clamped within the size of the pool on initialization. I had the same thought initially, but rounding it up caused unrelated tests to fail.

Reply:
But you agree it's slicing less than before? Note that slice() takes start and end arguments (inclusive), not start and length.

After thinking about this some more, I think that the rules/constraints should be:

1. `thisPool.used` is always a multiple of 8 (the pool itself is already suitably aligned), and

2. this particular slice() call should only slice multiples of 8 at offsets that are multiples of 8.

(1) implies (2) because (2) caches slices for reuse by (1).

(2) can be enforced by rounding `start + toRead` down.

Potential pitfall: rounding up or down can result in slices smaller than `kMinPoolSpace`. It should check that the slice's length `>= kMinPoolSpace` before pushing it into `poolFragments`.

Reply:
Yes it is slicing less than before. By rounding up the start position, we advance `8 - n_remainder` bytes past where we would have sliced. In this PR, `pool.used` (and `thisPool.used`) should always be a multiple of 8, which I believe addresses point (1).

With regards to (2), the byteLength of each slice as it is now isn't guaranteed to be a multiple of 8, though I think that's also a good idea. Performance-wise, it's best to align the `toRead` value to the processor's cache line size, but I left it as-is to keep the surface area of this PR to a minimum.

Good catch on the adjusted size causing the pool fragment size to dip below `kMinPoolSpace`, I'll update the PR with a fix here in a bit.